gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlSchedulerLeaseDeterminationStore.java

```diff
@@ -0,0 +1,207 @@
+    }
+    this.dagActionStore = dagActionStore;
+  }
+
+  @Override
+  public LeaseAttemptStatus attemptInsertAndGetPursuantTimestamp(String flowGroup, String flowName,
+      String flowExecutionId, FlowActionType flowActionType, long triggerTimeMillis)
+      throws IOException {
+    Timestamp triggerTimestamp = new Timestamp(triggerTimeMillis);
+    try (Connection connection = this.dataSource.getConnection();
+        PreparedStatement insertStatement = connection.prepareStatement(
+            String.format(ATTEMPT_INSERT_AND_GET_PURSUANT_TIMESTAMP_STATEMENT, tableName, tableName, epsilon, tableName,
+                epsilon))) {
+      int i = 0;
+      // Values to set in new row
+      insertStatement.setString(++i, flowGroup);
+      insertStatement.setString(++i, flowName);
+      insertStatement.setString(++i, flowExecutionId);
+      insertStatement.setString(++i, flowActionType.toString());
+      insertStatement.setTimestamp(++i, triggerTimestamp);
+      // Values to check if existing row matches
+      insertStatement.setString(++i, flowGroup);
+      insertStatement.setString(++i, flowName);
+      insertStatement.setString(++i, flowExecutionId);
+      insertStatement.setString(++i, flowActionType.toString());
+      insertStatement.setTimestamp(++i, triggerTimestamp);
+      // Values to make select statement to read row
+      insertStatement.setString(++i, flowGroup);
+      insertStatement.setString(++i, flowName);
+      insertStatement.setString(++i, flowExecutionId);
+      insertStatement.setString(++i, flowActionType.toString());
+      insertStatement.setTimestamp(++i, triggerTimestamp);
+      ResultSet resultSet = insertStatement.executeQuery();
+      connection.commit();
+
+      if (!resultSet.next()) {
+        resultSet.close();
+        throw new IOException(String.format("Unexpected error where no result returned while trying to obtain lease. "
+                + "This error indicates that no entry existed for trigger flow event for table %s flow group: %s, flow "
+                + "name: %s flow execution id: %s and trigger timestamp: %s when one should have been inserted",
+            tableName, flowGroup, flowName, flowExecutionId, triggerTimestamp));
+      }
+      // If a row was inserted, then we have obtained the lease
+      int rowsUpdated = resultSet.getInt(1);
+      if (rowsUpdated == 1) {
+        // If the pursuing flow launch has been persisted to the {@link DagActionStore} we have completed lease obtainment
+        this.dagActionStore.addDagAction(flowGroup, flowName, flowExecutionId, DagActionStore.DagActionValue.LAUNCH);
+        if (this.dagActionStore.exists(flowGroup, flowName, flowExecutionId, DagActionStore.DagActionValue.LAUNCH)) {
+          if (updatePursuantTimestamp(flowGroup, flowName, flowExecutionId, flowActionType, triggerTimestamp)) {
+            // TODO: potentially add metric here to count number of flows scheduled by each scheduler
```

Comment:
I think this would be an excellent metric also for the purpose of loadbalancing, maybe worth implementing directly